### ISSUE TITLE
[Deno] get code coverage working / solve rare crypto-related failures

### ISSUE THREAD

**Issue by cinnamon-bun** [earthstar-project/stone-soup#129]
## What's the problem you want solved?

`deno coverage` doesn't let you specify an import map, so we get this error from inside noble-ed25519:

```
error: Relative import path "crypto" not prefixed with / or ./ or ../
    at https://deno.land/x/ed25519@1.3.3/index.ts:6:24
```

To reproduce this, add to the Makefile:
```make
test-coverage:
	deno test --import-map=import_map.json --no-check=remote --config deno.json --coverage=coverage src

show-coverage:
	deno coverage coverage
```

## Is there a solution you'd like to recommend?

Maybe we can import noble-ed25519 through skypack or esm.sh to fix this?  It's also a pretty small package, we could just copy it and modify it.

**Comment by cinnamon-bun** [comment]
Importing it from `https://esm.sh/@noble/ed25519@1.4.0` fixes some things, but not code coverage.

It replaces `import "crypto"` with `import "https://deno.land/std@0.119.0/node/crypto.ts"`, and that means we don't need an import map anymore.  You can see this by running

```
deno info "https://esm.sh/@noble/ed25519@1.4.0"
```

**Comment by cinnamon-bun** [comment]
Also getting very rare test failures where an async op wasn't finished before the test finished, somewhere in the crypto code.  Maybe a result of forcing noble-ed25519 to use deno crypto instead of node crypto...

**Comment by cinnamon-bun** [comment]
noble-ed25519 is indeed calling crypto.web.subtle.digest here, and it is correctly awaiting the result:

https://github.com/paulmillr/noble-ed25519/blob/main/index.ts#L835-L839

```ts
    if (crypto.web) {
      const buffer = await crypto.web.subtle.digest('SHA-512', message.buffer);
      return new Uint8Array(buffer);
    } else if (crypto.node) {
      return Uint8Array.from(crypto.node.createHash('sha512').update(message).digest());
```

It should be using the node version though, as seen by running `deno info` (see above).  The [node version in deno.land/std/node/crypto.ts](https://deno.land/std@0.119.0/node/crypto.ts) is totally synchronous, there should be no chance of an async call happening...

**Comment by cinnamon-bun** [comment]
Here's the problem, noble-ed25519 is sniffing to see if it's in a browser or in Node:

https://github.com/paulmillr/noble-ed25519/blob/main/index.ts#L797-L802

```ts
// Global symbol available in browsers only. Ensure we do not depend on @types/dom
declare const self: Record<string, any> | undefined;
const crypto: { node?: any; web?: any } = {
  node: nodeCrypto,
  web: typeof self === 'object' && 'crypto' in self ? self.crypto : undefined,
};
```

Deno acts like a browser in this case, but we probably want it to use the node version of its code to avoid crashes.

**Comment by cinnamon-bun** [comment]
related PR: https://github.com/earthstar-project/stone-soup/pull/130

**Comment by sgwilym** [comment]
I did a little investigating into this this morning.

1. I created an [isolated test case](https://github.com/sgwilym/deno-async-crypto-repro/runs/4760166584?check_suite_focus=true) where a tiny copy of crypto driver _using the unmodified version of ed25519 served via deno.land_ was run against tests. After several attempts running this workflow on Github I could not get it to fail.
2. I created a new branch where I've undone your extraction of ed25519 into the codebase and imported the _unmodified version of ed25519 served via deno.land_ and repeated tests on Github workflows. I cannot get them to fail after several attempts.

This makes me think the following:

- Tests began failing after importing a transformed version of ed25519 via esm.sh (in [this commit](7e228b55f92f85e572f1a3116a4653deabdc261d)) and indeed looking at [this list](https://github.com/earthstar-project/stone-soup/commits/esm-first) you can see that's the case.

The good news is ed25519 works and there is no underlying async leak with Deno + linux.

---

This leaves the issue with coverage. `deno coverage` [does not yet support import maps](https://github.com/denoland/deno/issues/10395) and that is annoying.

I have another motivation for removing the need for an import map from the project: Deno does not automatically detect import maps and we would thus be requiring all downstream users of Earthstar to have an import map for ed25519's sake. I've made [an issue for that here](https://github.com/paulmillr/noble-ed25519/issues/40).

I have another idea to deal with this that I'm going to try now and report back on.

**Comment by sgwilym** [comment]
Please see the above PR, which I think has unravelled this mystery.

**Comment by sgwilym** [comment]
I think this is fixed by #132. The async crypto related failures are definitely fixed,  and generating coverage seems to work for me pretty consistently (the trick seems to be always running `make coverage` and not its subcommands independently).
